### ISSUE TITLE
etcd cluster instability

### ISSUE THREAD

**Issue by ndegory** [appcelerator-archive/amp#1642]
etcd is deployed with 3 replicas on 3 different nodes, which allows to lose a node and still get a functional cluster (raft consensus). Each replica uses a local Docker volume to persist its data.
When a worker node hosting an etcd replica disappears, Swarm launches a new task on one of the remaining nodes to reach the desired replica count. The new container is unable to start, since the data from the volume is already used by a running container.

The list of tasks of the etcd service after a worker node disappears. The missing node is the one with a node ID instead of a hostname. The new task shares the same host as an existing task.
```
$ docker service ps amp_etcd
ID                  NAME                IMAGE                      NODE                        DESIRED STATE       CURRENT STATE          ERROR               PORTS
10hvp87h5maq        amp_etcd.1          appcelerator/etcd:3.1.10   ip-192-168-0-52             Running             Running 37 hours ago
4lfes0v8xokt        amp_etcd.2          appcelerator/etcd:3.1.10   ip-192-168-0-52             Running             Running 2 hours ago
x6okutel1df2         \_ amp_etcd.2      appcelerator/etcd:3.1.10   9mindui4785i7ws9reoe67c75   Shutdown            Running 37 hours ago
ka8pyg28aima        amp_etcd.3          appcelerator/etcd:3.1.10   ip-192-168-16-228           Running             Running 37 hours ago
```

Logs from the new container.
```
$ docker logs <container_id>
2017-08-16 07:37:07.388689 W | etcdserver: another etcd process is running with the same data dir and holding the file lock.
2017-08-16 07:37:07.388712 W | etcdserver: waiting for it to exit before starting...
Error:  context deadline exceeded
```
Interestingly, this tasks doesn't error, and is waiting indefinitely. This doesn't allow Swarm to schedule it to another node.

Solution to be considered: 
- global service instead of replica count? That would put the constraint on the definition of the worker group, which is dangerous
- configure the deploy restart policy for the etcd service to make sure the new task is run after enough time to give time to the infrastructure to launch a new node. That's at best a guess and would reduce the risk, but won't fix it
- modify the appcelerator/etcd image to first check if another etcd container is running on same host, and exit if it's the case

**Comment by nikolai-derzhak-distillery** [comment]
We decided we need some watchdog service to remove old cluster members in separate container.

Probably it can be done as part of health check script also. But we have to make sure it will not bring issues with health status of containers. So dedicated container seems as porper way to implement it.

Draft of how it would work is next:

Watchdog will do periodic checks of service.$ETCD_SERVICE_NAME container IPs list for changes.

Once it detects IP that is gone (it means container is gone (healthchecks failed, whole node is down)) - it will remove it from member list of etcd raft using etcdctl CLI pointing to healthy node (?).

For new container and new IP in list - we need to make sure it will be added using CLI too. 
As it official way to replace failed node: https://coreos.com/etcd/docs/latest/op-guide/runtime-configuration.html#replace-a-failed-machine

That can be done in entrypoint run.sh script:
1.
```
$ etcdctl member add infra3 http://10.0.1.13:2380
added member 9bf1b35fc7761a23 to cluster
 ```
2. Start etcd with new IP list.

Things can get bad when we loose two container in the process of bring up new one.
But that is extreme case and will require manual threat.

**Comment by nikolai-derzhak-distillery** [comment]
Well cleanup can be done from new node before adding new member . Like we did for managers . I will validate that approach with quick POC.

**Comment by ndegory** [comment]
@nderzhak I think the two comments above are for #1643, right?
